### ISSUE TITLE
Add 'as' support to assertions

### ISSUE THREAD

**Issue by checketts** [kotlintest/kotlintest#475]
AssertJ has a ['describe' feature](http://joel-costigliola.github.io/assertj/assertj-core-features-highlight.html#describe-assertion) for assertions. 

```
assertThat(frodo.getAge()).as("check %s's age", frodo.getName()).isEqualTo(100);
//output: [check Frodo's age] expected:<100> but was:<33>
```

The current equivalent in kotlintest is:
```
frodo.age shouldBe 100
//output: expected:<100> but was:<33>
```

I would love to have the ability to add a context to a failure.  Perhaps like so:

```
"check ${frodo.name}'s age" { frodo.age shouldBe 100 }
// or to avoid adding invoke to string:
describe("check ${frodo.name}'s age") { frodo.age shouldBe 100 }
```

I have my own current extension functions that catch the `AssertionError` and wraps the context, but I would love to see it supported as a first class citizen in kotlintest. 

Is this a welcome feature? I would love to begin a PR.

**Comment by sksamuel** [comment]
There's something similar already called `withClue`.
If that works for you then great. If it's not powerful enough then a PR is most welcome to replace it / improve it.

**Comment by checketts** [comment]
`withClue` is a great start!

With `withClue("id") { actual.id shouldBe expected.id }`
```
java.lang.AssertionError: id org.opentest4j.AssertionFailedError: expected: "123" but was: "124"
Expected :"123" 
Actual   :"124"
```

I'll look into how hard it would be to get it to be:
```
java.lang.AssertionError: id expected: "123" but was: "124"
Expected :"123" 
Actual   :"124"
```

**Comment by Kerooker** [comment]
Closed as per #477
